File: lab-01-quizer/src/by/quackable/quizer/task_generator/PoolTaskGenerator.java
package by.quackable.quizer.task_generator;

import by.quackable.quizer.task.Task;

import java.util.List;
import java.util.Random;
import java.util.concurrent.ThreadLocalRandom;

public class PoolTaskGenerator implements TaskGenerator{
    PoolTaskGenerator(boolean allow_duplicate, Task... tasks) {
        this.tasks = List.of(tasks);
        this.allow_duplicate = allow_duplicate;
    }
    PoolTaskGenerator(boolean allow_duplicate, List<Task> tasks) {
        this.tasks = tasks;
        this.allow_duplicate = allow_duplicate;
    }

    @Override
    public Task generate() {
        if (tasks.isEmpty()) {
            throw new RuntimeException("Task list is empty");
        }
        Random random = ThreadLocalRandom.current();
        int index = random.nextInt(tasks.size());
        Task task = tasks.get(index);
        if (!allow_duplicate) {
            tasks.remove(index);
        }
        return task;
    }

    private final boolean allow_duplicate;
    private final List<Task> tasks;
}

File: lab-01-quizer/src/src/com.dranikpg.quizer/PoolTaskGenerator.kt
package com.dranikpg.quizer

import java.lang.RuntimeException

class PoolTaskGenerator private constructor(private val duplicates: Boolean,
                                            private val gs: MutableList<Task> = mutableListOf())
    : TaskGenerator {
    constructor(duplicates: Boolean, cgs: Collection<Task>) : this(duplicates, cgs.toMutableList())
    constructor(duplicates: Boolean, vararg vgs: Task) : this(duplicates, vgs.toMutableList())
    override fun generate(): Task = if (gs.size == 0) throw RuntimeException("No more tasks left")
        else (0 until gs.size).random().let { Pair(it, gs[it]) }
            .also { if (!duplicates) gs.removeAt(it.first) }
            .let { it.second }

}